File: frontend/src/types/retrievalPacket.ts
/**
 * Mirrors backend `app/schemas/retrieval_packet.py` and `app/schemas/retrieval.py`
 * (Phase 4 File Clerk + Phase 7 context recording on the packet).
 * Used for typed consumption of `POST /retrieve` — no mock payloads.
 */

export type PacketType = "retrieval_packet";

export type IntentType =
  | "explain"
  | "compare"
  | "locate"
  | "summarize"
  | "debug"
  | "recommend"
  | "unknown";

export type AnswerMode =
  | "answer_with_evidence"
  | "answer_with_caveats"
  | "ask_clarification"
  | "not_enough_evidence"
  | "conflicting_evidence"
  | "unsupported";

export interface InterpretedIntent {
  intent_type: IntentType;
  confidence: number;
  notes: string[];
}

export interface SelectedSemanticIndex {
  semantic_index_id: string;
  meaning: string;
  score: number;
  selection_reason: string;
}

export interface GraphPathPacket {
  start_node_id: string;
  nodes: string[];
  edges: string[];
  depth: number;
}

export interface EvidenceUnitPacket {
  evidence_unit_id: string;
  artifact_id: string;
  modality: string;
  content_type: string;
  source_fidelity: string;
  text?: string | null;
  location?: Record<string, unknown> | null;
  selection_reason: string;
  confidence?: number | null;
}

export interface EvidenceLanguageAggregateRow {
  language: string;
  evidence_unit_count: number;
  average_confidence?: number | null;
  min_confidence?: number | null;
  max_confidence?: number | null;
}

/** Phase 7 — derived from selected evidence `metadata_json` only; not translation; not evidence. */
export interface RetrievalLanguageContext {
  version: number;
  source: string;
  query_language?: string | null;
  evidence_languages: EvidenceLanguageAggregateRow[];
  primary_evidence_language?: string | null;
  distinct_evidence_language_count: number;
  evidence_units_without_language_metadata_count: number;
  warnings: string[];
}

export interface AlternativeInterpretation {
  if_user_meant: string;
  suggested_semantic_indexes: string[];
  reason: string;
}

export interface ContextBudgetSummary {
  max_evidence_units: number;
  selected_evidence_units: number;
  pruned_evidence_units: number;
  pruning_reasons: string[];
  max_graph_paths?: number | null;
  max_selected_indexes?: number | null;
}

/** Optional `POST /retrieve` request metadata (Phase 7); validated server-side; not used for routing. */
export interface ActorContext {
  actor_id?: string | null;
  actor_type?: string | null;
  role?: string | null;
  expertise_level?: string | null;
  preferred_language?: string | null;
  purpose?: string | null;
  metadata?: Record<string, unknown> | null;
}

/** Subset of options exposed in Query Playground; backend fills other defaults. */
export interface RetrieveOptionsPayload {
  max_evidence_units: number;
  max_graph_depth: number;
  include_alternatives: boolean;
}

export interface RetrieveRequestPayload {
  question: string;
  options?: RetrieveOptionsPayload | null;
  /** Optional; omitted when unset. Recording-only on the packet in the current baseline. */
  actor_context?: ActorContext | null;
}

export type ActorContextInfluence = "none" | "recorded_only_no_route_boost_applied";

/** Phase 7 — how request actor context was reflected on the packet. */
export interface PacketActorContextRecording {
  used: boolean;
  source: string;
  recorded_context?: ActorContext | null;
  influence: ActorContextInfluence;
  warnings: string[];
}

export interface RetrievalPacket {
  packet_type: PacketType;
  question: string;
  interpreted_intent: InterpretedIntent;
  selected_indexes: SelectedSemanticIndex[];
  graph_paths: GraphPathPacket[];
  evidence_units: EvidenceUnitPacket[];
  alternative_interpretations: AlternativeInterpretation[];
  context_budget: ContextBudgetSummary;
  warnings: string[];
  confidence: number;
  answer_mode: AnswerMode;
  language_context?: RetrievalLanguageContext | null;
  actor_context?: PacketActorContextRecording | null;
}

File: backend/app/schemas/retrieval_packet.py
"""RetrievalPacket JSON contracts: Phase 4 File Clerk plus Phase 7 context extensions."""

from __future__ import annotations

from typing import Any, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.schemas.retrieval import ActorContext

PacketType = Literal["retrieval_packet"]

ActorContextInfluence = Literal["none", "recorded_only_no_route_boost_applied"]

IntentType = Literal["explain", "compare", "locate", "summarize", "debug", "recommend", "unknown"]

AnswerMode = Literal[
    "answer_with_evidence",
    "answer_with_caveats",
    "ask_clarification",
    "not_enough_evidence",
    "conflicting_evidence",
    "unsupported",
]

class InterpretedIntent(BaseModel):
    """Deterministic query intent classification output."""

    intent_type: IntentType
    confidence: float = Field(ge=0.0, le=1.0)
    notes: list[str] = Field(default_factory=list)


class SelectedSemanticIndex(BaseModel):
    """A semantic index chosen for graph entry and traversal."""

    semantic_index_id: str
    meaning: str
    score: float
    selection_reason: str


class GraphPathPacket(BaseModel):
    """Bounded traversal snapshot attached to the packet."""

    start_node_id: str
    nodes: list[str]
    edges: list[str]
    depth: int


class EvidenceUnitPacket(BaseModel):
    """Evidence grounded via graph support links, included in the packet."""

    evidence_unit_id: str
    artifact_id: str
    modality: str
    content_type: str
    source_fidelity: str
    text: str | None = None
    location: dict[str, Any] | None = None
    selection_reason: str
    confidence: float | None = None


class EvidenceLanguageAggregateRow(BaseModel):
    """Per-language rollups over selected evidence metadata (Phase 7)."""

    language: str
    evidence_unit_count: int
    average_confidence: float | None = None
    min_confidence: float | None = None
    max_confidence: float | None = None


class RetrievalLanguageContext(BaseModel):
    """Language routing metadata derived from selected evidence ``metadata_json`` only.

    Not source truth; no detection or translation in Phase 7 Slice 7F.
    """

    version: Literal[1] = 1
    source: Literal["selected_evidence_metadata"] = "selected_evidence_metadata"
    query_language: str | None = None
    evidence_languages: list[EvidenceLanguageAggregateRow] = Field(default_factory=list)
    primary_evidence_language: str | None = None
    distinct_evidence_language_count: int = 0
    evidence_units_without_language_metadata_count: int = 0
    warnings: list[str] = Field(default_factory=list)


class AlternativeInterpretation(BaseModel):
    """Explicit ambiguity / alternate meaning routing hints."""

    if_user_meant: str
    suggested_semantic_indexes: list[str]
    reason: str


class ContextBudgetSummary(BaseModel):
    """Visible context budgeting decisions."""

    max_evidence_units: int
    selected_evidence_units: int
    pruned_evidence_units: int
    pruning_reasons: list[str]
    max_graph_paths: int | None = None
    max_selected_indexes: int | None = None


class RetrieveOptions(BaseModel):
    """Optional tuning for a single retrieval request."""

    max_evidence_units: int = Field(default=8, ge=1, le=64)
    max_graph_depth: int = Field(default=1, ge=0, le=5)
    max_graph_paths: int = Field(default=3, ge=1, le=32)
    max_selected_indexes: int = Field(default=3, ge=1, le=16)
    include_alternatives: bool = True
    max_tokens_estimate: int | None = Field(default=None, ge=1, le=100_000)


class RetrieveRequest(BaseModel):
    """HTTP body for `POST /retrieve`."""

    question: str
    options: RetrieveOptions | None = None
    actor_context: ActorContext | None = None


class PacketActorContextRecording(BaseModel):
    """How request ``ActorContext`` was reflected on the packet (Phase 7 Slice 7H).

    Recording-only: never evidence and must not imply retrieval influence beyond ``influence``.
    """

    model_config = ConfigDict(extra="forbid")

    used: bool
    source: Literal["request_actor_context"] = "request_actor_context"
    recorded_context: ActorContext | None = None
    influence: ActorContextInfluence
    warnings: list[str] = Field(default_factory=list)


class RetrievalPacket(BaseModel):
    """Structured evidence routing output from the File Clerk."""

    packet_type: PacketType = "retrieval_packet"
    question: str
    interpreted_intent: InterpretedIntent
    selected_indexes: list[SelectedSemanticIndex]
    graph_paths: list[GraphPathPacket]
    evidence_units: list[EvidenceUnitPacket]
    alternative_interpretations: list[AlternativeInterpretation]
    context_budget: ContextBudgetSummary
    warnings: list[str]
    confidence: float = Field(ge=0.0, le=1.0)
    answer_mode: AnswerMode
    language_context: RetrievalLanguageContext | None = None
    actor_context: PacketActorContextRecording | None = None
